lib/primitives/mtx.js

```diff
@@ -1688,13 +1711,58 @@ class CoinSelector {
    */
 
   fund() {
-    // Ensure all preferred inputs first.
-    if (this.inputs.size > 0) {
-      const coins = [];
+    this.resolveInputs();
+    this.resolvePreferred();
+
+    if (this.isFull())
+      return;
+
+    while (this.index < this.coins.length) {
+      const coin = this.coins[this.index++];
```

Comment:
Is it possible that a coin would be added twice? If, for example, `this.coins[1]` was actually spent by one of the preferred or existing inputs, could that coin be mistakenly added a second time to fund the rest of the transaction? I actually am not even sure `mtx.check()` would catch a double-spend like this, we might need an explicit test